### ISSUE TITLE
[Proposal] Card database changes 

### ISSUE THREAD

**Issue by ctrlaltca** [Cockatrice/Cockatrice#2545]
Since the introduction of the "id" field on mtgjson, i started working on a branch that uses that information to discriminate between cards with the same name.
The initial idea was to make people able to select card arts from the edition they prefer, but it turned out to be a quite invading change that could introduce more benefits.
I'm working on this change on this branch: https://github.com/Cockatrice/Cockatrice/compare/master...ctrlaltca:mtgjson_next

Currently this is the list of changes:

About the card hash:
- the "card_hash" is a new exposed server and client feature
- card hashes are populated with the card id coming from mtgjson
- each card in the card database is identified by its hash; cards with a different hash will be shown as different cards, even if they have the same name. This means that in the deck editor you'll find all the cards variations from all the sets, each one with its different text, art, etc.
- as a consequence, each card in the card database is now associated to only 1 set; previously, each card maintained a list of all the sets where it appeared.
- the card hash is passed in all the server-client communications involving cards (serverinfo_card, command_create_token, event_create_token, event_move_card , ..) so the other players can know exactly the card edition that you are playing (and see the same card art as you)
- old clients not supporting the feature will just use the card name as before.

About the xml database format:
- the xml format of cards.xml, tokens.xml, etc has been changed to add the hash field
- the xml format of cards.xml, tokens.xml, etc has been changed to move all the mtg-specific fields as generic named properties; the card database will load all the defined properties, so implementing new games will be easier.
- the card database version has been incremented from 3 to 4, so old xml files are not compatible.
- the token database url has been changed to ensure old clients downloads the old xml version and new clients downloads the new xml version

Currently the branch needs to be updated with the latest changes from `master` and tested.

**Comment by tritoch** [comment]
Sounds great, closes #2330 

How are custom sets handled?

Can a user choose to override the per-set card images and continue to use the mUID lookup card or a 'default' card image?

**Comment by Psithief** [comment]
That idea of having one line per printing in the deck editor is just throwing usability out the window. You'll have to do a UI overhaul to address that, or people will want to downgrade immediately.

---

So you want me to generate hashes for new tokens? You're turning updating the tokens into a process that can't be done by hand, **but** that's the only way it _can be done_.

This also raises questions about how people will create their own custom card xmls.

**Comment by tritoch** [comment]
> So you want me to generate hashes for new tokens? You're turning updating the tokens into a process that can't be done by hand, but that's the only way it can be done.

Why not just move the card name with X amount of spaces after it to the hash field?  Or make the hash TokenName# where # is the number of spaces you're using currently?

**Comment by ctrlaltca** [comment]
> How are custom sets handled?
Like before, just the xml format has changed.

> Can a user choose to override the per-set card images and continue to use the mUID lookup card or a 'default' card image?
i don't understand the question, the mUID is already per-set.

> That idea of having one line per printing in the deck editor is just throwing usability out the window. You'll have to do a UI overhaul to address that, or people will want to downgrade immediately.

Agreed; my idea is to only show the card from the "first enabled set" and add small [+] button on the left of the card ow that shows the sublist of card editions. This is not present in the code yet.

> So you want me to generate hashes for new tokens? You're turning updating the tokens into a process that can't be done by hand, but that's the only way it can be done.

Card hashes are not mandatory, so you are not forced into adding them in tokens.xml.
Anyway, adding card hashes could avoid the use of the "whitespace hack", aka #473 

> Why not just move the card name with X amount of spaces after it to the hash field? Or make the hash TokenName# where # is the number of spaces you're using currently?

A good proposal is to drop the spaces entirely, but to be able to discriminate between cards with the same name. Generating a hash (random or hashing card name + properties) for each token is a possibility.

Please remember this is a proposal about an idea i'm working on, i'm not going to push it down your throat, maybe it will turn out to be a bad idea and i'll just discard it.

**Comment by Psithief** [comment]
The xml format is just so hard to work with. It's over-normalising the data, and I don't know if anyone is actually, genuinely committed to this ephemeral idea of making Cockatrice a platform for multiple card games.

If there's less emphasis on that, then the next obvious xml format is something like:

```xml
<cards>
  <card>
    <name>card name</name>
    <set>
      <name>set name 1</name>
      <hash>1092b8c39c01f823a9f01</hash>
      <muid>3213123</muid>
      <setnumber>145</setnumber>
      <rarity>Mythic Rare</rarity>
    </set>
    <set>
      <name>set name 2</name>
      <hash>c01f823a9f011092b8c39</hash>
      <muid>593853</muid>
      <setnumber>102</setnumber>
      <rarity>Rare</rarity>
    </set>
    <manacost>1G</manacost>
    <colors>G</colors>
    <cmc>2</cmc>
    <text>Not a tarmogoyf.</text>
    <pt>*/*-1</pt>
  </card>
</cards>
```

**Comment by Daenyth** [comment]
We can make tooling to support. I'm also open to stuff other than XML.
SQLite or json of some kind maybe.

**Comment by ZeldaZach** [comment]
If we choose to use json, we can simply append to the current mtgjson files and it'll be easier to maintain

**Comment by tooomm** [comment]
The new uniqueID coming with mtgjson v4 could just replace the hash?
https://github.com/mtgjson/mtgjson4#mtgjson

**Comment by Daenyth** [comment]
I believe ctrl is basing the hash on mtgjsonv4's uniqueID

**Comment by Jinxstands4Jinx** [comment]
I still think we should be able to see what and all sets reprinted cards are in. it sounds like this will make cards not show then thy are in multiple sets.

**Comment by Daenyth** [comment]
@Jinxstands4Jinx the opposite - this proposal gives cockatrice *more* power to deal with multiple printings

**Comment by Jinxstands4Jinx** [comment]
yes But will we still be able to know how many printings cards have had, and what the other sets were?

**Comment by Nightfirecat** [comment]
I presume that information will still be displayed when you select the card.

**Comment by tooomm** [comment]
>yes But will we still be able to know how many printings cards have had, and what the other sets were?

That's a question of presentation. I'm sure we don't wont to completely drop that information. ;)
@ctrlaltca even suggested a solution in one of his post:
>Agreed; my idea is to only show the card from the "first enabled set" and add small [+] button on the left of the card ow that shows the sublist of card editions. This is not present in the code yet.

Also, the `Card info` widget in deck editor or the middle click on a card name in chat could show all available printings for example.

**Comment by tritoch** [comment]
We had an issue with all available printings being too verbose for things like basic lands.  Neither here nor there in regards to this topic though.

**Comment by ZeldaZach** [comment]
New brainstorming session for MTGJSONv4 integration:

Each card should contain as much information as possible that could be useful

1) Name
2) Printings
2a) Rarity
2b) MultiverseID
2c) ScryfallID
2d) Card number in set
2e) Set Code
2f) Face of card (front/back) -- Will only need "back" (can use MTGJSON.side for this info)
2g) MTGJSON UUID
3) color(s)
4) mana cost
5) cmc
6) type
7) where to put card (table row)
8) oracle text
9) legalities
9a) \<legality status="">FORMAT\</legality> where status can be BANNED, RESTRICTED, or LEGAL
10) power/toughness and/or loyalty
11) frame version (useful for alternative formats)
12) foreign data -- So users can localize the cards/text
12a) \<foreign muId="" text="" name="">LANGUAGE\</foreign>
13) artist of the card -- new filter field

This information is _all_ obtainable via MTGJSONv4, plus more. I have omitted fields that are probably not relevant (flavor text, for example), but if others want to chime in suggestions Full docs are at https://mtgjson.com/docs.html

**Comment by ebbit1q** [comment]
Would adding that many fields slow down loading of the card database on startup even more? The entire client just hangs for a while until it's done.

**Comment by tooomm** [comment]
There have been thoughts to support a SQL database over XML.

We already highly advice users to use their own separate files for custom sets etc. and don't change our cards.xml file at all. It's also written at the top inside the xml file. Because of that I think it probably wouldn't matter too much if it would be a real database and the user couldn't edit it that easily anymore. Maybe that even helps with some issues and broken databases. :)
For custom stuff the XML support should still be around!

**Comment by Psithief** [comment]
> For custom stuff the XML support should still be around!

I agree, taking a human-readable format away from people making their own databases makes it a nightmare for support.

**Comment by olegshtch** [comment]
Most other TCG have some unique "id" while card names could be same.

**Comment by ZeldaZach** [comment]
Related to #3961 and #2330. I'd like to discuss more info about a potential database upgrade to support bigger and better things.
